File: blueprints/monitoring.py
import random
from flask import Blueprint, render_template, jsonify
from flask_login import login_required
from models.device import Device

monitoring_bp = Blueprint('monitoring', __name__, url_prefix='/monitoring')


@monitoring_bp.route('/<int:device_id>')
@login_required
def detail(device_id):
    """Monitoring detail page for a specific device."""
    device = Device.query.get_or_404(device_id)
    return render_template('monitoring/detail.html', device=device)


@monitoring_bp.route('/<int:device_id>/interface')
@login_required
def interface_view(device_id):
    """Interface (port) visualization page."""
    device = Device.query.get_or_404(device_id)
    return render_template('monitoring/interface.html', device=device)


@monitoring_bp.route('/api/<int:device_id>/realtime')
@login_required
def realtime_data(device_id):
    """
    Real-time monitoring data endpoint.
    Returns simulated data for scaffolding; replace with actual
    RouterOS API / Ansible data in production.
    """
    device = Device.query.get_or_404(device_id)

    # Simulated real-time data for demo purposes
    data = {
        'device_id': device.id,
        'device_name': device.name,
        'uptime': device.uptime or '10d 4h 30m',
        'routeros_version': device.routeros_version or '7.15.2',
        'model': device.model or 'RB750GR3',
        'cpu_load': random.randint(10, 60),
        'memory_used': random.randint(150, 400),
        'memory_total': 512,
        'temperature': random.randint(25, 50),
        'disk_used': random.randint(20, 80),
        'disk_total': 128,
        'total_users': random.randint(1, 20),
        'active_connections': random.randint(5, 50),
        'interfaces': [
            {
                'name': 'ether1',
                'tx_rate': f'{random.randint(100, 9999)} kbps',
                'rx_rate': f'{random.randint(100, 9999)} kbps',
                'tx_bytes': f'{random.randint(1, 500)} MB',
                'rx_bytes': f'{random.randint(1, 500)} MB',
            },
            {
                'name': 'ether2',
                'tx_rate': f'{random.randint(100, 9999)} kbps',
                'rx_rate': f'{random.randint(100, 9999)} kbps',
                'tx_bytes': f'{random.randint(1, 500)} MB',
                'rx_bytes': f'{random.randint(1, 500)} MB',
            },
            {
                'name': 'ether3',
                'tx_rate': f'{random.randint(100, 9999)} kbps',
                'rx_rate': f'{random.randint(100, 9999)} kbps',
                'tx_bytes': f'{random.randint(1, 500)} MB',
                'rx_bytes': f'{random.randint(1, 500)} MB',
            },
            {
                'name': 'ether4',
                'tx_rate': f'{random.randint(100, 9999)} kbps',
                'rx_rate': f'{random.randint(100, 9999)} kbps',
                'tx_bytes': f'{random.randint(1, 500)} MB',
                'rx_bytes': f'{random.randint(1, 500)} MB',
            },
            {
                'name': 'ether5',
                'tx_rate': f'{random.randint(100, 9999)} kbps',
                'rx_rate': f'{random.randint(100, 9999)} kbps',
                'tx_bytes': f'{random.randint(1, 500)} MB',
                'rx_bytes': f'{random.randint(1, 500)} MB',
            },
        ],
        'port_status': {
            'wan': 'connected',
            'ether1': 'connected',
            'ether2': 'connected',
            'ether3': 'disabled',
            'ether4': 'disconnected',
            'ether5': 'disconnected',
        }
    }
    return jsonify(data)

File: static/js/monitoring.js
/**
 * NETSIBLE — Monitoring real-time polling
 */
document.addEventListener('DOMContentLoaded', () => {
    const container = document.getElementById('monitoring-container');
    if (!container) return;
    const deviceId = container.dataset.deviceId;
    if (!deviceId) return;

    function update() {
        fetch(`/monitoring/api/${deviceId}/realtime`)
            .then(r => r.json())
            .then(d => {
                setText('mon-uptime', d.uptime);
                setText('mon-ros', d.routeros_version);
                setText('mon-model', d.model);
                setText('mon-cpu', d.cpu_load + '%');
                setText('mon-mem', d.memory_used + ' MB / ' + d.memory_total + ' MB');
                setText('mon-temp', d.temperature + '°C');
                setText('mon-disk', d.disk_used + ' MB / ' + d.disk_total + ' MB');
                setText('mon-users', d.total_users);
                setText('mon-conn', d.active_connections);
                // Update interface table
                const tbody = document.getElementById('intf-table-body');
                if (tbody && d.interfaces) {
                    tbody.innerHTML = d.interfaces.map(i => `<tr>
                        <td>${i.name}</td><td>${i.tx_rate}</td><td>${i.rx_rate}</td>
                        <td>${i.tx_bytes}</td><td>${i.rx_bytes}</td></tr>`).join('');
                }
            }).catch(() => {});
    }

    function setText(id, val) {
        const el = document.getElementById(id);
        if (el) el.textContent = val;
    }

    update();
    setInterval(update, 5000);
});
